Automated Docker Image Publish to Docker Hub and GitHub Packages at Releases

HarryHung opened the pull request:
This action will run on publish of releases automatically.

It will first run the existing build-to-test action. If successful, it will then build and publish the image to both Docker Hub and GitHub Packages. 

It requies Action Secrets `DOCKER_USERNAME` and `DOCKER_PASSWORD` set up in Settings of the repo. **[We don't have it yet!]**
(`GITHUB_TOKEN` is automatically generated by the action)

HarryHung committed:
Create GH action for automated deployment

HarryHung committed:
Merge pull request #1 from HarryHung/feature/automated-deployment

Create GH action for automated deployment

HarryHung committed:
Correct Docker Image tag

HarryHung committed:
Allow reusing this workflow

HarryHung committed:
Build to test before publishing image

HarryHung committed:
Publish to Docker Hub in addition to GitHub Packages

HarryHung committed:
Re-add build-to-test before publish

HarryHung committed:
Fix images publish

Oliver-Lorenz-dev commented:
Looks good Harry, cool to get an auto-deploy CI running!
I guess we need to hold off on merging until we get the credentials?

HarryHung commented:
Yes, we need to wait for the credentials. I have tested it on mine in my fork and the Action itself is working. 
In theory if we don't trigger the Action by publishing a release, the time of merging should not matter. 
But let's wait just in case.

HarryHung committed:
Build to app stage only for publish image

HarryHung committed:
Re-adding correct Docker Image tag

HarryHung commented:
Added two fixes.

- Forgot to skip the `test` stage in the final image, fixed. 
- Accidentally removed the line to ensure correct image tag (no leading `v`), fixed.
